File: Python/list/My Calendar II.py
class MyCalendarTwo:

    def __init__(self): 
        self.events=[]
        self.overlaps=[]

    def book(self, start: int, end: int) -> bool:
        # Step 1: Check if the new event overlaps with any of the previously double-booked intervals
        for ov in self.overlaps:
            overlapStart = ov[0]
            overlapEnd=ov[1]
            if overlapStart<end and start<overlapEnd:
                return False
        # Step 2: Check for new overlaps with existing events and update 'overlaps' list     
        for ev in self.events:
            eventStart = ev[0]
            eventEnd = ev[1]
            if start<eventEnd and eventStart<end:
                overlapStart = max(start, eventStart)
                overlapEnd= min(end,eventEnd)
                self.overlaps.append((overlapStart, overlapEnd))
        # Step 3: No triple booking found, so add the new event to 'events' list
        self.events.append((start,end))
        return True

# Your MyCalendarTwo object will be instantiated and called as such:
# obj = MyCalendarTwo()
# param_1 = obj.book(start,end)

File: CPP/map/My Calendar II.cpp
class MyCalendarTwo {
public:
    vector<pair<int,int>>events;
    vector<pair<int,int>>overlaps;
    MyCalendarTwo() {
        
    }
    bool book(int start, int end) {
        //check for new events causes a triple booking
        for(auto ov:overlaps){
            int overlapStart = ov.first;
            int overlapEnd= ov.second;
            if(overlapStart <end && start <overlapEnd){
                // If the new event overlaps with any range in the 'overlaps', it would cause a triple booking
                return false; 
            }
        }
        //check for overlaps with all existing events
        for(auto ev:events){
            int startEvent = ev.first;
            int endEvent =ev.second;
            if(start<endEvent && startEvent<end){
                int overlapStart = max(start,startEvent);
                int overlapEnd = min(end,endEvent);
                overlaps.push_back({overlapStart,overlapEnd});
            }
        }
        // If no triple booking is detected, add the event to the list of events
        events.push_back({start, end});
        return true;
    }
};

/**
 * Your MyCalendarTwo object will be instantiated and called as such:
 * MyCalendarTwo* obj = new MyCalendarTwo();
 * bool param_1 = obj->book(start,end);
 */
